File: bukkit/elysium-players-bukkit/src/main/java/com/seventh_root/elysium/players/bukkit/BukkitPlayerProvider.java
package com.seventh_root.elysium.players.bukkit;

import com.seventh_root.elysium.api.player.PlayerProvider;
import com.seventh_root.elysium.core.database.Table;
import com.seventh_root.elysium.players.bukkit.database.table.BukkitPlayerTable;
import org.bukkit.OfflinePlayer;

public class BukkitPlayerProvider implements PlayerProvider<BukkitPlayer> {

    private final ElysiumPlayersBukkit plugin;

    public BukkitPlayerProvider(ElysiumPlayersBukkit plugin) {
        this.plugin = plugin;
    }

    @Override
    public BukkitPlayer getPlayer(int id) {
        return plugin.getCore().getDatabase().getTable(BukkitPlayer.class).get(id);
    }

    public BukkitPlayer getPlayer(OfflinePlayer bukkitPlayer) {
        Table<BukkitPlayer> table = plugin.getCore().getDatabase().getTable(BukkitPlayer.class);
        if (table instanceof BukkitPlayerTable) {
            BukkitPlayerTable bukkitPlayerTable = (BukkitPlayerTable) table;
            BukkitPlayer player = bukkitPlayerTable.get(bukkitPlayer);
            if (player == null) {
                player = new BukkitPlayer(bukkitPlayer);
                addPlayer(player);
            }
            return player;
        }
        return null;
    }

    @Override
    public void addPlayer(BukkitPlayer player) {
        plugin.getCore().getDatabase().getTable(BukkitPlayer.class).insert(player);
    }

    @Override
    public void removePlayer(BukkitPlayer player) {
        plugin.getCore().getDatabase().getTable(BukkitPlayer.class).delete(player);
    }

}

File: bukkit/elysium-players-bukkit/src/main/kotlin/com/seventh_root/elysium/players/bukkit/BukkitPlayerProvider.kt
package com.seventh_root.elysium.players.bukkit

import com.seventh_root.elysium.api.player.PlayerProvider
import com.seventh_root.elysium.core.database.Table
import com.seventh_root.elysium.players.bukkit.database.table.BukkitPlayerTable
import org.bukkit.OfflinePlayer

class BukkitPlayerProvider(private val plugin: ElysiumPlayersBukkit) : PlayerProvider<BukkitPlayer> {

    override fun getPlayer(id: Int): BukkitPlayer? {
        return plugin.core.database.getTable(BukkitPlayer::class.java)!![id]
    }

    fun getPlayer(bukkitPlayer: OfflinePlayer): BukkitPlayer {
        val table = plugin.core.database.getTable(BukkitPlayer::class.java)
        var player = (table as BukkitPlayerTable)[bukkitPlayer]
        if (player == null) {
            player = BukkitPlayer(bukkitPlayer)
            addPlayer(player)
        }
        return player
    }

    override fun addPlayer(player: BukkitPlayer) {
        val bukkitPlayerTable: Table<BukkitPlayer> = plugin.core.database.getTable(BukkitPlayer::class.java) as Table<BukkitPlayer>
        bukkitPlayerTable.insert(player)
    }

    override fun removePlayer(player: BukkitPlayer) {
        plugin.core.database.getTable(BukkitPlayer::class.java)!!.delete(player)
    }

}
